# Ejercicio 1
## a) 
Mostrar que $\alpha$ es raíz primitiva. 

Cuando sea $p = 163$ y $\alpha = 3$

In [128]:
from numpy import * #andy
#Se crea un arreglo que va del 0,162
a = zeros(163)

print("Arreglo de 0,162")
print(a)

for i in range (1,163):
    res = 3**i % 163
    a[res] = 1
    #print("Iteración ",i ,"resultado", res)
a

Arreglo de 0,162
[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


array([0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

Con esto vemos que se obtenieron todos los enteros de [1,162] puesto que se lleno el arreglo con un uno a excepción de la posición $0$ lo cual verifica que $\alpha$ es raíz primitiva

b) De todas las raíces primitivas

In [129]:
for i in range (1,163):#andy
    res = 3**i % 163
    print("3^",i ,"mod 163 = ", res)

3^ 1 mod 163 =  3
3^ 2 mod 163 =  9
3^ 3 mod 163 =  27
3^ 4 mod 163 =  81
3^ 5 mod 163 =  80
3^ 6 mod 163 =  77
3^ 7 mod 163 =  68
3^ 8 mod 163 =  41
3^ 9 mod 163 =  123
3^ 10 mod 163 =  43
3^ 11 mod 163 =  129
3^ 12 mod 163 =  61
3^ 13 mod 163 =  20
3^ 14 mod 163 =  60
3^ 15 mod 163 =  17
3^ 16 mod 163 =  51
3^ 17 mod 163 =  153
3^ 18 mod 163 =  133
3^ 19 mod 163 =  73
3^ 20 mod 163 =  56
3^ 21 mod 163 =  5
3^ 22 mod 163 =  15
3^ 23 mod 163 =  45
3^ 24 mod 163 =  135
3^ 25 mod 163 =  79
3^ 26 mod 163 =  74
3^ 27 mod 163 =  59
3^ 28 mod 163 =  14
3^ 29 mod 163 =  42
3^ 30 mod 163 =  126
3^ 31 mod 163 =  52
3^ 32 mod 163 =  156
3^ 33 mod 163 =  142
3^ 34 mod 163 =  100
3^ 35 mod 163 =  137
3^ 36 mod 163 =  85
3^ 37 mod 163 =  92
3^ 38 mod 163 =  113
3^ 39 mod 163 =  13
3^ 40 mod 163 =  39
3^ 41 mod 163 =  117
3^ 42 mod 163 =  25
3^ 43 mod 163 =  75
3^ 44 mod 163 =  62
3^ 45 mod 163 =  23
3^ 46 mod 163 =  69
3^ 47 mod 163 =  44
3^ 48 mod 163 =  132
3^ 49 mod 163 =  70
3^ 50 mod 163 =  47

# Ejercicio 2
## a)
Mediante el algoritmo de paso grande paso chico encontrar el logaritmo de $19$ base $\alpha = 3$ explicando detalladamente como se llega al resultado.

Es decir, hay que calcular $\log_{3}(19)$ tenemos entonces, $\beta = 19$ y por el ejercicio anterior que $Z^{*}_
{163}$ el orden del grupo es $162$ y $m = \lceil \sqrt{162} \rceil = 13$  entonces $0 \leq j < 13$ lo cual nos dice que hay que calcular $3^j$ donde $j$ toma valores en el intervalo $[0,12]$, pero eso lo hicimos para el ejercicio anterior por lo que ocupamos la siguiente funcion:

In [7]:
from numpy import *

def cal(x):
    a = zeros(x)
    for i in range (0,x):
        res = 3**i % 163
        a[i]=res
    return a

In [9]:
arr = cal(12)
arr

array([  1.,   3.,   9.,  27.,  81.,  80.,  77.,  68.,  41., 123.,  43.,
       129.])

Calculamos entonces $\alpha^{-1} = 3^{-1} \text{ mod } 163 = 109$ este calculo fue a "mano", como son números pequeños el inverso lo calculamos de la siguiente manera, el modulo es $163$ y $163*2 = 326$ y por otro lado tenemos que $3*109 = 327$ lo que nos da $1$ en el modulo, por lo tanto $109$ es el inverso multiplicativo de $3$.

Teniendo lo anterior tenemos que calcular $$\alpha^{-1m} = \alpha^{-m} = 109^{13} \text{mod } 163$$

In [17]:
r1 = 109**13 % 163
r1

106

Entonces $\alpha^{-1m} = \alpha^{-m} = 109^{13} \text{mod } 163 = 106$

$$\alpha = 3$$ y $$\alpha^{-mi} = 106^i$$ y $$\beta = 19$$
Esto por propiedades de los exponentes. Con esto podemos usar la siguiente función para calcular el logaritmo requerido

In [24]:
""" 
Función que genera hasta que beta por alfa a la m*i sea igual a alfa
alfa: entero
    es el número generador, dicho de otra manera, base del calculo del logaritmo
beta: entero
    Numero al calcular el logaritmo
alfa_prima: entero
    el número que se elevara al número de iteración.
modulo: entero
    modulo en el que se esta trabajando
"""
def generando(alfa,beta,alfa_prima, modulo):
    i=-1
    r=0
    while(r!=alfa):
        i=i+1
        r = beta * (alfa_prima)**i % modulo  
    return i

In [25]:
generando(3,19,106,163)

30

Con la función *generando* con parametros $3,19,106,163$ obtenemos que en la iteración $30$ se obtuvo el generador lo cual significa lo siguiente: 
$$\beta\alpha^{-30m} = 3 = \alpha^1$$
De antemano sabemos que $m = 13$ entonces
$$\beta\alpha^{-30*13} = 3 = \alpha^1$$
$$\beta\alpha^{-390} = \alpha^1$$
Por lo cual obtenemos que $$\beta\alpha^{-390} = \alpha^1$$ entonces $$\beta = \alpha^{391}$$

Entonces $\log_{3} 19 = 391$

Verificación de lo obtenido. 

In [26]:
t = 3**391 % 163
t

19

## b)

Con el algoritmo de Pohling Hellman calcular $\log_{3}(19)$ desarrollando paso a paso 


Tenemos la siguiente información
$Z^{*}_{163}$ 
$$ p = 163, n = 162, \alpha = 3, \beta = 19$$ 
Primero obtenemos la factorización de $n$ que es $$n = 162 = 2 * 3 * 3 * 3 * 3$$ y es única ahora calculamos 

$X_{i} \equiv X \text{mod } 2$ 

Paara $i=1$ calculamos entonces

\begin{align} 
\alpha^{\prime} & = \alpha^{n/2} \text{mod }p \\
& = \alpha^{162/2} \text{mod } 163 \\
& = (3)^{81} \text{mod } 163 \\
& = 162
\end{align}

Además calculamos *beta* prima

\begin{align} 
\beta^{\prime} & = \beta^{n/2} \text{mod } 163\\
 & = 19^{81} \text{mod } 163\\
 & = 162\\
\end{align}

$X_1 = log_{162}(162) = 1$ Ya que el logaritmos de un número n en base n es igual a 1 entonces $X_{1} \equiv 1 mod 2$ 


In [37]:
3**81 % 163

162

In [38]:
19**81 % 163

162

$i=2$ entonces $X_2 = l_0 + l_1 * 3 + l_2 * 3^2 + l_3 * 3^3$

\begin{align} 
X_2 & = l_0 + l_1 * 3 + l_2 * 3^2 + l_3 * 3^3\\
   & = l_0 + l_1 * 3 + l_2 * 9 + l_3 * 27\\
\end{align}

donde $X_2 = X mod 3^4$

ahora $q=3$ y $e=4$

asignamos $\gamma = 1$ y $l_{-1} = 0$ 

Calculamos

\begin{align} 
\alpha^{\prime} & = \alpha^{162/3}  \\
& = \alpha^{54} \text{mod } 163 \\
& = 58\\
\end{align}$$

Calculamos $l_j$ donde $j\in \{0,1,2,3\}$

$j=0$ calculamos 

\begin{align} 
\gamma & = 1*\alpha^{(l_{j-1})*(q^{j-1})}\\
       & = 1*\alpha^{(0)(3^{-1})}\\
       & = 1*\alpha^0\\
       & = 1(3) mod 163\\
       & = 3\\
\end{align}

Conociendo el valor de Gamma podemos calcular el valor de beta prima.

\begin{align} 
\beta^{\prime} & = (19(3)^{-1})^{n/q^{j+1}} mod 163\\
        & = (19(3)^{-1})^{162/3^1} mod 163\\
        & = (19*109)^{54} mod 163\\
        & = 1\\
\end{align}

El inverso de $3$ modulo $163$ ya lo habiamos calculado para el inciso a.

$l_0 = log_3(1) = 162$ entonces $l_0 = 162$ ya que $3^{162}mod163 = 1$

Ahora $j=1$ calculamos $\gamma$

\begin{align}
\gamma & = 1\alpha^{l_0*q^0}\\
    & =1(\alpha)^{(162)(3)^0}\\
    & = 1(3)^{162}\\
\end{align}

Teniendo el valor de $\alpha$ podemos calcular $\beta$

\begin{align}
\beta & = (\beta\gamma^{-1})^{n/q^{j+1}}\\
    & = (19((3^{162})^{-1})^{18} mod 163\\
\end{align}




In [5]:
3**54 % 163

58

In [4]:
(19*109)**54 % 163


1

Tendriamos que calcula el inverso de $3^{162}$ pero es un número demasiado grande además de que creemos que no calculamos bien los exponentes, consultamos el libro de Menezes pero no nos quedo claro esa parte del algoritmo, sin embargo, había que seguir calculando las $l_i$ hasta poder realizar la suma de cada una y así obteniamos otra ecuación de congruencias las que había que resolver, en este caso como son dos, había que usar sustitución y así obtener el logaritmo discreto. No es mucho, bueno si es bastante, no todo, pero es trabajo honesto. 

## Ejericicio 4

Sea $n=23999$ tenemos que tomar una $a$ en el intervalo $[2,23997]$

In [130]:
def simboloLegendre(mod,queda): #Andy
    for i in range (1,23999):
        res = i**2 % mod
        if res == queda:
            #print("Iteración ",i ,"resultado", res)
            return 1
    return -1
    #Si devuelve 1, entonces simbolo de legendre es 1
    #Si devuelve -1, entonces simbolo de legendre es 1

In [113]:
def simboloJacobi():
    for i in range (2,23997):
        m103 = simboloLegendre(103,i)
        m233 = simboloLegendre(233,i)
        Jaco = m103 * m233
        if Jaco == -1:
            print("La a es:",i)
            break

In [114]:
simboloJacobi()

La a es: 17


Calculamos $r$

In [122]:
17**11999 % 23999

17668

Solo faltaria verificar que $r \ncong s \text{ mod } n$ donde  $r=17668, n = 23999$ y $s$ siendo el símbolo de Jacobi con $a=17$ y $s=-1$ entonces $17668 \ncong -1 \text{ mod } 23999$ por lo que $23999$ no es primo

In [119]:
def func(mod,queda):
    for i in range (1,23999):
        res = i**2 % mod
        if res == queda:
            return 1
    return -1

In [120]:
func(103,17)


1

In [121]:
func(233,17)

-1

# Ejercicio 3 Descifrar el mesaje desde el criptosistema ElGammal

G(163,3,19) entonces $p=163, \alpha = 3, \alpha ^a = 19$ y tenemos mensajes de tipo $(\gamma,\delta)$, para descifrar hay que calcular $\gamma^{p-1-a}$ 

Calculando el valor de $a$ en la ecuación $3^a$ mod $163 = 19$ con fuerza bruta


In [4]:
for i in range (1,163):
    res = 3**i % 163
    if res == 19:
        print("Iteración ",i ,"resultado", res)

Iteración  67 resultado 19


Obteniendo así que el valor de $a = 67$

Ahora que tenemos el valor de $a$ podemos descifrar cada uno de los mensajes, puesto que son de la forma: 

$$ M = \gamma^{p-1-a} * \delta \text{ mod } p $$ y a continuación tenemos una función que lo hará, pues ya conocemos todos los valores para poder hacer el calculo.

## Función para calcular $r$ en una ecuación de la forma

 $$b^a mod p = r$$


Es decir, que esta función nos sirve para descifrar en el sistema de ElGammal

In [6]:
"""
Función llamada mensajeXY que calcula el valor en número del mensaje codificado en ASCII del 65 al 90
donde la ñ se toma como n 

El mensaje esta codificado y representado por una pareja ordenada de la forma (x,y)

x: entero
primer número de la codificación del mensaje

y: entero
segundo número de la codificación del mensaje

p: entero
número primo que se tiene como llave pública

a: entero
llave privada del intercambio de mensajes

"""
def mensajeXY(x,y,p,a):
    m = x**(p-1-a) * y % p
    return m


Para el caso especial del ejercicio 3 de la tarea 2 tenemos la siguiente función.


In [27]:
def mensajeXY3(x,y):
    m = (x**(95)) * y  % 163
    return m

In [31]:
mensajeXY3(7,27)

69

In [32]:
mensajeXY3(7,75)

83

In [33]:
mensajeXY3(7,125)

84

In [34]:
mensajeXY3(7,38)

79

In [35]:
mensajeXY3(5,108)

69

In [36]:
mensajeXY3(5,137)

83

In [37]:
mensajeXY3(7,12)

85

In [38]:
mensajeXY3(7,151)

78

In [39]:
mensajeXY3(7,153)

65

In [40]:
mensajeXY3(7,25)

82

In [41]:
mensajeXY3(11,80)

69

In [42]:
mensajeXY3(7,90)

67

In [43]:
mensajeXY3(5,152)

79

In [44]:
mensajeXY3(7,101)

77

In [45]:
mensajeXY3(7,88)

80

In [46]:
mensajeXY3(2,5)

69

In [47]:
mensajeXY3(5,115)

78

In [48]:
mensajeXY3(11,23)

83

In [49]:
mensajeXY3(5,123)

65

Con la tabla de ASCII obtenemos el mensaje descifrado

### El mensaje descifrado es: Esto es una recompensa

# Ejercicio 6 

Sea $n = 23999$ tenemos que es 11-homogenea ya que $2*3*5*7*11 = 2310$ ya que con el siguiente primo tendriamos que $2*3*5*7*11*13 \geq 23999$ el algoritmo de Pollard rho-1 (Segun el Menezes) nos dice que si encontramos que $d = mcd(a,n) \geq 2$ la salida del algoritmo es $d$ dado que en el ejercicio 4 donde obtuvimos la factorización de $n=23999 = 103 * 233$ sabemos que ya tenemos un numeró $a$ con el cual se cumple esta condición, calculada a continuación:  

In [124]:
def mcd(a,b): #Andy
    while(b>0):
        divisor = a % b
        a = b
        b = divisor
    return a

Sea $a=103$ calculamos el mcd de $n = 23999$ y $ a = 103 $

Para encontrar todas los factores bastaria iterar a en el intervalo $[2,23998]$ pero dado lo hecho en el ejercicio 4 sabemos que los números $103, 233$ son lo que cumplen la condición de que $d=mcd(a,n) \geq 2$

In [125]:
mcd(103,23999)

103

In [126]:
mcd(233,23999)

233

In [127]:
103*233

23999